HIVE-27269: Create multiple MatchTracker when using VectorFastHashMapContainer

ngsg opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://cwiki.apache.org/confluence/display/Hive/HowToContribute
  2. Ensure that you have created an issue on the Hive project JIRA: https://issues.apache.org/jira/projects/HIVE/summary
  3. Ensure you have added or run the appropriate tests for your PR: 
  4. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP]HIVE-XXXXX:  Your PR title ...'.
  5. Be sure to keep the PR description updated to reflect all changes.
  6. Please write your PR title to summarize what this PR proposes.
  7. If possible, provide a concise example to reproduce the issue for a faster review.

-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

This patch changes VectorMapJoinFastHashMapContainer to create multiple MatchTracker, each of which tracks distinct HashMap in the HashMapContainer.
In addition, it fixes fixed size array allocation problem in 3 classes and adjusts the number of partitioned HashTable when it is not a power of 2.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Current VectorMapJoinFastHashMapContainers have 1 MatchTracker and multiple HashMaps. MatchTracker tracks MATCHED entries by marking the index of HashTable entry in a bitset. Since multiple HashMaps use the same MatchTracker, if one of the HashTables marks the index as MATCHED, then the other HashMap's entry on the same index is also treated as MATCHED regardless of actual join result. This leads to missing non-matched join results of FullOuterJoin.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description, screenshot and/or a reproducable example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Hive versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No

### Is the change a dependency upgrade?
<!--
If yes, please attach a file with output from mvn dependency:tree to validate a complete upgrade of dependency.
If no, write 'No'.
-->
No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
I added unit tests for Long, String, and MultiKey VectorizedFastHashMapContainer. Also, I ran TPC-DS query97 on 100GB dataset for verifying the patch.

ngsg commented:
@ayushtkn , I added a qfile which reproduces the issue using long and string key. Could you please review this PR again? Thank you.

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastHashTableLoader.java:
```diff
@@ -106,9 +107,21 @@ public void init(ExecMapperContext context, MapredContext mrContext,
+
+      int adjustedValue = Integer.highestOneBit(initialValue);
+      if (initialValue != adjustedValue) {
+        LOG.info("Adjust the number of HT-loading-threads to {}. (Previous value: {})",
```
what's the use of that, should we auto-tune?
````
 this.numLoadThreads=adjustedValue?
````

ngsg replied:
Yes, I intended auto-tuning, and the code is wrong. Thanks for pointing it out.

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastLongHashMapContainer.java:
```diff
@@ -109,10 +109,13 @@ public static class NonMatchedLongHashMapIterator extends VectorMapJoinFastNonMa
         VectorMapJoinFastLongHashMap[] vectorMapJoinFastLongHashMaps, int numThreads) {
       super(matchTracker);
+
+      assert matchTracker.getIsPartitioned();
```
assertions should not be used in production code

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/fast/VectorMapJoinFastLongHashMapContainer.java:
```diff
@@ -185,8 +188,14 @@ public JoinUtil.JoinResult lookup(long key, VectorMapJoinHashMapResult hashMapRe
 
   public JoinUtil.JoinResult lookup(long key, VectorMapJoinHashMapResult hashMapResult,
       MatchTracker matchTracker) {
+    assert matchTracker == null || matchTracker.getIsPartitioned();
```
same, assertions should not be used in production code

deniskuzZ commented:
@ngsg, isn't it the same for VectorFastHashSetContainer?

ngsg commented:
@deniskuzZ , if my memory serves me correctly, this problem happens when Hive uses MatchTracker. Only FullOuterJoin uses MatchTracker, and FOJ always uses HashMap. So I think this issue is irrelevant to HashSet and HashMultiSet. I'll check it again and leave you a comment if there is something wrong.

ngsg commented:
@deniskuzZ , I fixed numHTTLoadhreads adjustment code and removed assertions from the patch. Also I have checked that HashSet/HashMultiSet containers do not have this issue. Could you please review the changes?

deniskuzZ commented:
> @deniskuzZ , I fixed numHTTLoadhreads adjustment code and removed assertions from the patch. Also I have checked that HashSet/HashMultiSet containers do not have this issue. Could you please review the changes?

Could you please explain why that issue is not applicable to HashSet/HashMultiSet containers? 
Based on the description in a PR I would assume that all classes introduced in `HIVE-25583: Support parallel load for HastTables` are affected.
````
Current VectorMapJoinFastHashMapContainers have 1 MatchTracker and multiple HashMaps. MatchTracker tracks MATCHED entries by marking the index of HashTable entry in a bitset. Since multiple HashMaps use the same MatchTracker, if one of the HashTables marks the index as MATCHED, then the other HashMap's entry on the same index is also treated as MATCHED regardless of actual join result.````

ngsg commented:
As I mentioned above, MatchTracker is used only by FOJ, and FOJ does not use HashSet/HashMultiSet container.

In current implementation, `VectorMapJoinOperator` sets `MatchTracker` by VectorMapJoinOuterGenerateResultOperator.fullOuterHashTableSetup(), which is called only by `VMJFullOuter{Long/String/MultiKey}Operator`. According to the `Vectorizer.specializeMapJoinOperator()`, these operators are chosen when `joinType` is `JoinDesc.FULL_OUTER_JOIN` and they always use `HashTableKind.HASH_MAP`. Therefore, this issue does not happen when using HashSet/HashMultiSet.

--- The remaining part explains about current fast VectorMapJoinHashTable implementation in order to help understanding the patch.

HIVE-25583 introduces parallel HT loading by creating multiple HashTables and putting them into a single HashTableContainer. (e.g. VectorMapJoinFastLongHashMapContainer has an array of VectorMapJoinFastLongHashMap.) The number of HashTables in a container is equal to the number of HT-loading threads, and a HashTable for the given KV pair(or Key in case of HashSet/HashMultiSet) is determined by the following formula: `(Hash of key) & (# of HT threads - 1)`.
```
e.g.
key hash = 0 -> use HT #0
key hash = 7, numHTThreads = 2 -> use HT #1
key hash = 7, numHTThreads = 4 -> use HT #3
```

Every HashTable has a long array which represents entries of HashTable. HashTable inserts a KV pair in 3 steps:
1. write KV pair on somewhere, usually a buffer
2. compute a long value that refer to the position of the written KV pair.
3. find an empty HashTable entry and save the long value
The last step is done by quadratic hash probing starting from the following index: `(Hash of key) & (length of the array)`.
```
e.g.
Insert key hash = 0 -> use HT #0, use slot #0
Insert key hash = 0 -> use HT #0, use slot #1 (slot #0 is already filled)
Insert key hash = 0 -> use HT #0, use slot #3 (slot #0, #1 are already filled)
```

MatchTracker tracks the slot index. MapJoinOperator may have a MatchTracker, and if it has one, it passes MatchTracker to HashTable when performing lookup. If HashTable finds matched element, then it records the slot index of matched element to MatchTracker.
```
e.g.
HashTable entries: {5, 0, 0, 6} where 0 represents an empty entry.
Lookup 1 -> MatchTracker: no slots matched
Lookup 5 -> MatchTracker: slot #0 matched
Lookup 6 -> MatchTracker: slot #0, #3 matched
```

Current implementation uses a single MatchTracker for all HashTables, which makes false-negative errors when listing non-matched elements.
```
e.g.
HT #0: {1, 5, 0, 0}, HT #1: {0, 2, 0, 0}, 0 represents an empty entry
Lookup 1 -> MatchTracker: slot #0 matched
Lookup 2 -> MatchTracker: slot #0, #1 matched
HT #0, non-matched entries according to the above MatchTracker: empty <- wrong result, should be slot #1
```

deniskuzZ commented:
> slot #0 matched



> As I mentioned above, MatchTracker is used only by FOJ, and FOJ does not use HashSet/HashMultiSet container.
> 
> In current implementation, `VectorMapJoinOperator` sets `MatchTracker` by VectorMapJoinOuterGenerateResultOperator.fullOuterHashTableSetup(), which is called only by `VMJFullOuter{Long/String/MultiKey}Operator`. According to the `Vectorizer.specializeMapJoinOperator()`, these operators are chosen when `joinType` is `JoinDesc.FULL_OUTER_JOIN` and they always use `HashTableKind.HASH_MAP`. Therefore, this issue does not happen when using HashSet/HashMultiSet.
> 
> --- The remaining part explains about current fast VectorMapJoinHashTable implementation in order to help understanding the patch.
> 
> [HIVE-25583](https://issues.apache.org/jira/browse/HIVE-25583) introduces parallel HT loading by creating multiple HashTables and putting them into a single HashTableContainer. (e.g. VectorMapJoinFastLongHashMapContainer has an array of VectorMapJoinFastLongHashMap.) The number of HashTables in a container is equal to the number of HT-loading threads, and a HashTable for the given KV pair(or Key in case of HashSet/HashMultiSet) is determined by the following formula: `(Hash of key) & (# of HT threads - 1)`.
> 
> ```
> e.g.
> key hash = 0 -> use HT #0
> key hash = 7, numHTThreads = 2 -> use HT #1
> key hash = 7, numHTThreads = 4 -> use HT #3
> ```
> 
> Every HashTable has a long array which represents entries of HashTable. HashTable inserts a KV pair in 3 steps:
> 
> 1. write KV pair on somewhere, usually a buffer
> 2. compute a long value that refer to the position of the written KV pair.
> 3. find an empty HashTable entry and save the long value
>    The last step is done by quadratic hash probing starting from the following index: `(Hash of key) & (length of the array)`.
> 
> ```
> e.g.
> Insert key hash = 0 -> use HT #0, use slot #0
> Insert key hash = 0 -> use HT #0, use slot #1 (slot #0 is already filled)
> Insert key hash = 0 -> use HT #0, use slot #3 (slot #0, #1 are already filled)
> ```
> 
> MatchTracker tracks the slot index. MapJoinOperator may have a MatchTracker, and if it has one, it passes MatchTracker to HashTable when performing lookup. If HashTable finds matched element, then it records the slot index of matched element to MatchTracker.
> 
> ```
> e.g.
> HashTable entries: {5, 0, 0, 6} where 0 represents an empty entry.
> Lookup 1 -> MatchTracker: no slots matched
> Lookup 5 -> MatchTracker: slot #0 matched
> Lookup 6 -> MatchTracker: slot #0, #3 matched
> ```
> 
> Current implementation uses a single MatchTracker for all HashTables, which makes false-negative errors when listing non-matched elements.
> 
> ```
> e.g.
> HT #0: {1, 5, 0, 0}, HT #1: {0, 2, 0, 0}, 0 represents an empty entry
> Lookup 1 -> MatchTracker: slot #0 matched
> Lookup 2 -> MatchTracker: slot #0, #1 matched
> HT #0, non-matched entries according to the above MatchTracker: empty <- wrong result, should be slot #1
> ```

@ngsg, thanks again for the detailed explanation! I understand the problem, I was just worried that we'll get exactly the same issues with HashSet/HashMultiSet maybe not with FOJ, but somewhere else. Anyhow, I'll merge this PR once we get a green build.
